docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,430 @@
+- Rollback protection: (Requires authentication) CMAC is stored in a designated rollback protected storage (also of KVStore type) when set and compared with the stored one at get.
+- Write once: Key can only be stored once and can't be removed. 
+
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce and external RBP key. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
+
+![SecureStore Record](./SecureStore_record.jpg)
+
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP Keys are generated using a 32-bit incremental counter, which is incremented whenever we have a new key. Start value of the counter is achieved by iterating over the rollback protection KVStore (at initialization) and getting the maximal value. 
+
+ 
+# Detailed design
+
+![SecureStore Class Hierarchy](./SecureStore_class_hierarchy.jpg)
+
+Functionality, as defined by KVStore, includes the following:
+- Initialization & reset
+- Core actions: get, set & remove
+- Incremental set actions
+- Iterator actions
+
+### Class header
+
+SecureStore has the following header:
+
+```C++
+class SecureStore : KVStore {
+
+public:
+    SecureStore(KVStore *underlying_kv, KVStore *rbp_kv);
+    virtual ~SecureStore();
+    	 
+    // Initialization and formatting
+    virtual int init();
+    virtual int deinit();
+    virtual int reset();
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *pattern=NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
+    virtual int iterator_close(iterator_t it);
+    
+private:
+    Mutex _mutex;
+    KVStore *_underlying_kv;
+    KVStore *_rbp_kv;
+    void *_entropy;
```

Comment:
is void * the right type in mbedtls?

Reply:
Yes - casted in CPP file to avoid exposure of type in header (a known C++ problem).